files/rsyslog/rsyslog.sh

```diff
@@ -228,6 +232,83 @@ fi
 # make sure there is not one left over from a previous run
 rm -f /var/run/rsyslogd.pid
 
+# Generate a config file to keep undefined fields
+keep_fields_lookup=${RSYSLOG_WORKDIRECTORY}/keep_fields.json
+keep_fields_conf=${RSYSLOG_WORKDIRECTORY}/10-keep_fields.conf
+rm -f $keep_fields_lookup $keep_fields_conf
+
+cat > $keep_fields_lookup << EOF
+{ "version" : 1,
+  "nomatch" : "unknown",
+  "type" : "string",
+  "table" : [
+EOF
+
+cat > $keep_fields_conf << EOF
+lookup_table(name="keep_fields" file="$keep_fields_lookup")
+
+template(name="debugTemplate" type="list") {
+    constant(value="\$!: ")
+    property(name="\$!")
+    constant(value="\\n")
+}
+
+template(name="firstFieldTemplate" type="list") {
+    constant(value="\\"") property(name="\$.ii!key") constant(value="\\":\\"")
+    property(name="\$.ii!value") constant(value="\\"")
+}
+
+template(name="addFieldTemplate" type="list") {
+    property(name="\$.undef")
+    constant(value=",\\"") property(name="\$.ii!key") constant(value="\\":\\"")
+    property(name="\$.ii!value") constant(value="\\"")
+}
+
+template(name="makeJsonTemplate" type="list") {
+    constant(value="{") property(name="\$.undef") constant(value="}")
+}
+
+# Moving undefined fields to \$!undefined
+if (strlen(\`echo \$CDM_USE_UNDEFINED\`) > 0) and (\`echo \$CDM_USE_UNDEFINED\` == "true") then {
+    set \$.field = "unknown";
+    set \$.firsttime = "true";
+    foreach (\$.ii in \$.all) do {
+        set \$.field = lookup("keep_fields", \$.ii!key);
+        if (strlen(\$.field) > 0) and (\$.field == "unknown") then {
+            if \$.firsttime == "true" then {
+                set \$.undef = exec_template('firstFieldTemplate');
+                set \$.firsttime = "false";
+            } else {
+                set \$.undef = exec_template('addFieldTemplate');
+            }
+        }
+    }
+    set \$!undefined = exec_template('makeJsonTemplate');
+}
+EOF
+
+for field in $( echo $DEFAULT_KEEP_FIELDS | sed -e "s/,/ /g" )
+do
+    cat >> $keep_fields_conf << EOF
+if strlen(\$.all!$field) > 0 then {
+    set \$!$field = \$.all!$field;
+}
+EOF
+    echo "    {\"index\": \"$field\", \"value\": \"true\"}," >> $keep_fields_lookup
+done
+
+for field in $( echo $EXTRA_KEEP_FIELDS | sed -e "s/,/ /g" )
+do
+    cat >> $keep_fields_conf << EOF
+if strlen(\$.all!$field) > 0 then {
+    set \$!$field = \$.all!$field;
+}
```

Comment:
After removing the `$field` from `$.all`, `$.all` will contain _only_ the "undefined" fields - the fields _not_ specified by `DEFAULT_KEEP_FIELDS` or `EXTRA_KEEP_FIELDS`.  Then, "move" `$.all` to `$!$undefined`:
`set \$!$undefined = \$.all;`
I don't know if this will work - https://www.rsyslog.com/doc/v8-stable/rainerscript/variable_property_types.html does not explicitly say what will happen if you assign an object from `$.` to what you want to be a new object in `$!`.

Also, I think we are reaching the limits of what we can do with generated config code + rscript.  If we need to do something much more complex than this, we will probably have to look into using https://www.rsyslog.com/doc/v8-stable/configuration/modules/mmexternal.html.  This is similar to what the new rsyslog_exporter is doing: https://github.com/openshift/cluster-logging-operator/blob/master/files/rsyslog/10-viaq-modules.conf#L34-L38 - it is using `omprog` to execute an external program to serve the prometheus metrics.  The external program is a go program: https://github.com/openshift/origin-aggregated-logging/tree/master/rsyslog/go/src/github.com/soundcloud/rsyslog_exporter

We may have to do something similar for complex record manipulation - write a go program like the `rsyslog_exporter` to parse/process/format a new JSON record.  I think go would be a good balance between ease of coding and performance.  If the performance is bad we could always write our own C rsyslog plugin but I'd rather avoid that.

Reply:
Thanks, @richm!  It looks much better (and simpler)!  Let me go this way a little more.